source/nodejs/adaptivecards/src/markdown-formatter.ts

```diff
@@ -0,0 +1,182 @@
+					MarkdownFormatter.MD_FORMATTER_CONFIG[i] = MarkdownFormatter.regexArray[j];
+					MarkdownFormatter.regexArray.splice(j, 1);
+					continue;					
+				}
+			}
+		}
+        MarkdownFormatter.regexArray = MarkdownFormatter.MD_FORMATTER_CONFIG.concat(MarkdownFormatter.regexArray);
+        
+    }
+    
+
+	constructor(text:string) {
+        this.text = text;
+       
+		for (var i = 0; i < MarkdownFormatter.MD_FORMATTER_CONFIG.length; i++) {
+			let patternArray = MarkdownFormatter.MD_FORMATTER_CONFIG[i].pattern;
+			let patternType = MarkdownFormatter.MD_FORMATTER_CONFIG[i].patternType;
+            let groups = MarkdownFormatter.MD_FORMATTER_CONFIG[i].groups;
+            let pattern = patternArray[0];
+            
+            if(patternType === "custom"){                
+                pattern = patternArray[0];
+            }
+            if(patternType === "start-end"){ 
+                pattern = patternArray[0].replace(MarkdownFormatter.SPECIAL_CHAR_REGEX, "\\$&") + '[\\s]+((.*?)[' + patternArray[1].replace(MarkdownFormatter.SPECIAL_CHAR_REGEX, "\\$&")+ "](?="+patternArray[1].replace(MarkdownFormatter.SPECIAL_CHAR_REGEX, '\\$&')+"[\\s]+)|(.*?)$)";
+            }else if(patternType == "symmetric"){
+                pattern = patternArray[0].replace(MarkdownFormatter.SPECIAL_CHAR_REGEX, "\\$&") + '(.*?)' + patternArray[0].replace(MarkdownFormatter.SPECIAL_CHAR_REGEX, "\\$&");    
+            }else if(patternType == "asymmetric"){                
+                pattern = patternArray[0].replace(MarkdownFormatter.SPECIAL_CHAR_REGEX, "\\$&");    
+                let regexForm = '';
+                let p = MarkdownFormatter.MD_FORMATTER_CONFIG[i].pattern[0];
+                for (var j = 0; j < p.length; j++) {
+                    regexForm = regexForm + "\\" + p[j];
+                }
+                // create all group regex
+				let part = regexForm.length / groups;
+				let regex = "";
+				for (var j = 0; j < groups; j++) {
+					let group = regexForm.substring(part * j, part * (j + 1));
+					let firstHalf = group.substring(0, group.length / 2);
+					let secondHalf = group.substring(group.length / 2, group.length);
+					let middle = (j < groups / 2) ? group.substring(0, group.length / 2) : group.substring(group.length / 2, group.length);
+					regex = regex + firstHalf + '([^' + middle + ']+)' + secondHalf;
+				}  
+				pattern = regex;
+			}
+            this.patterns[i] = new RegExp(pattern, 'gim');
+			this.styles[i] = MarkdownFormatter.MD_FORMATTER_CONFIG[i].styles;
+			this.styleTypes[i] = MarkdownFormatter.MD_FORMATTER_CONFIG[i].type;
+			this.patternTypes[i] = patternType;
+		}
+	}
+
+	render() {
+        
+		for (var i = 0; i <= this.styleTypes.length - 1; i++) {		
+			this.text = this.parseText(this.text, this.styleTypes[i], this.styles[i], this.patterns[i], this.patternTypes[i]);
+		}
+
+		return this.text;
+	}
+
+	parseText = (text, type, styles, pattern, patternType) => {
```

Comment:
- Should be `private`
- All parameters must be typed
- Return value must be typed
- Why this inconsistent method declaration syntax? Should be
```typescript
private parseText(text, ...): string { ... }
```
- Given that this method is called like this:
```typescript
this.parseText(this.text, this.styleTypes[i], this.styles[i], this.patterns[i], this.patternTypes[i]);
```
why does it need to be passed all these parameters? They can all go and be replaced with a simple `index: number`.